### PR TITLE
feat: support ZETA with V2 architecture

### LINKED ISSUE #358: Implement `depositAndRevert` using ZETA tokens in addition to ZRC20
Currently the `depositAndRevert` logic is implemented at the protocol level for ZETA reverts.

For consistency in with V2 contracts and simplifying the protocol logic, this logic should be added in the protocol contracts using `depositAndRevert`

Low priority as ZETA is not enabled for now

### LINKED ISSUE #394: Support back ZETA for gateway
Support back withdrawing and depositing using gateway on ZEVM and EVM.

Part of https://github.com/zeta-chain/node/issues/2739

Opening this issue to track the part in the code to comment back (use ctrl+f with this issue link)

### LINKED ISSUE #496: Write tests for ZETA support in the gateway test scripts
Write a comprehensive test suite for ZETA support within the contracts v2 architecture.

### PR DESCRIPTION
**Pull request by s2imonovic**
## Changes
- Restored support for ZETA token withdrawals and deposits via gateway on both ZEVM and EVM
- Implemented depositAndRevert functionality for ZETA tokens (previously only available for ZRC20)

Closes #394
Closes #358
Closes #496

## Summary by CodeRabbit

- **New Features**
  - Introduced support for depositing ZETA tokens, enabling new deposit and revert operations on ZEVM.
  - Added a public function to deposit ZETA tokens and trigger a callback on a specified contract.

- **Improvements**
  - Enhanced withdrawal logic for ZETA tokens, including validation and event emission.
  - Updated function and error names for improved clarity and consistency.
  - Gas fees are now burned during ZRC20 withdrawals for increased transparency.

- **Bug Fixes**
  - Corrected event emission and balance assertions in test cases for ZETA operations.

- **Tests**
  - Added comprehensive tests for new deposit and revert functionality.
  - Expanded tests to cover gas fee burning and edge cases in withdrawal scenarios.

- **Chores**
  - Removed deprecated error messages and updated function names for better maintainability.

### CODE REVIEW

File: contracts/evm/GatewayEVM.sol [hunk 1]
```diff
@@ -382,18 +382,13 @@ contract GatewayEVM is
...
-            // IERC20(token).safeTransferFrom(from, address(this), amount);
-            // // approve connector to handle tokens depending on connector version (eg. lock or burn)
-            // if (!IERC20(token).approve(zetaConnector, amount)) revert ApprovalFailed();
-            // // send tokens to connector
-            // ZetaConnectorBase(zetaConnector).receiveTokens(amount);
+            // transfer to connector
+            // transfer amount to gateway
+            IERC20(token).safeTransferFrom(from, address(this), amount);
+            // approve connector to handle tokens depending on connector version (eg. lock or burn)
+            if (!IERC20(token).approve(zetaConnector, amount)) revert ApprovalFailed(token, zetaConnector);
```
[the remark this answers is not in the document]
Comment (s2imonovic) [reply contracts/evm/GatewayEVM.sol:389]: there is no `safeApprove()`, you probably meant the `forceApprove()` function from the SafeERC20 library. I’d rather do that separately as a contracts improvement instead of adding it to this PR

File: contracts/zevm/GatewayZEVM.sol [hunk 2]
```diff
@@ -149,8 +149,12 @@ contract GatewayZEVM is
     function _withdrawZRC20WithGasLimit(uint256 amount, address zrc20, uint256 gasLimit) private returns (uint256) {
         (address gasZRC20, uint256 gasFee) = IZRC20(zrc20).withdrawGasFeeWithGasLimit(gasLimit);
 
-        if (!_safeTransferFrom(gasZRC20, msg.sender, PROTOCOL_ADDRESS, gasFee)) {
-            revert GasFeeTransferFailed(gasZRC20, PROTOCOL_ADDRESS, gasFee);
+        if (!_safeTransferFrom(gasZRC20, msg.sender, address(this), gasFee)) {
```
[the remark this answers is not in the document]
Comment (s2imonovic) [reply contracts/zevm/GatewayZEVM.sol:152]: Maybe I’d leave it as it is for now, since in case of an error during transferFrom it reverts with a different error, and it’s only used in two places for now.

**Review by fadeev — APPROVED** [review]

File: contracts/zevm/GatewayZEVM.sol [hunk 3]
```diff
@@ -149,8 +149,12 @@ contract GatewayZEVM is
...
+            revert GasFeeTransferFailed(gasZRC20, address(this), gasFee);
+        }
+
+        if (!_safeBurn(gasZRC20, gasFee)) {
+            revert ZRC20BurnFailed(gasZRC20, gasFee);
```
Comment (lumtis) [inline contracts/zevm/GatewayZEVM.sol:157]: This is https://github.com/zeta-chain/protocol-contracts/issues/480?

I would it would make more sense to have a dedicated PR for this one since a bit different from the rest so we can keep cleaner atomicity and commit history

File: contracts/zevm/GatewayZEVM.sol [hunk 4]
```diff
@@ -300,31 +300,27 @@ contract GatewayZEVM is
...
-        // if (message.length + revertOptions.revertMessage.length > MAX_MESSAGE_SIZE) revert MessageSizeExceeded();
-
-        // _transferZETA(amount, PROTOCOL_ADDRESS);
-        // emit WithdrawnAndCalled(
-        //     msg.sender, chainId, receiver, address(zetaToken), amount, 0, 0, message, callOptions, revertOptions
-        // );
+        if (receiver.length == 0) revert ZeroAddress();
+        if (amount == 0) revert InsufficientZetaAmount();
+        if (callOptions.gasLimit < MIN_GAS_LIMIT) revert InsufficientGasLimit();
+        if (message.length + revertOptions.revertMessage.length > MAX_MESSAGE_SIZE) {
```
Comment (lumtis) [inline contracts/zevm/GatewayZEVM.sol:316]: There are checks that can maybe be factorized across other withdraw function, if so I suggest to open an issue and add a TODO in the comment

File: contracts/evm/GatewayEVM.sol [hunk 5]
```diff
@@ -382,18 +382,12 @@ contract GatewayEVM is
...
-            // // approve connector to handle tokens depending on connector version (eg. lock or burn)
-            // if (!IERC20(token).approve(zetaConnector, amount)) revert ApprovalFailed();
-            // // send tokens to connector
-            // ZetaConnectorBase(zetaConnector).receiveTokens(amount);
+            // transfer amount to gateway
+            IERC20(token).safeTransferFrom(from, address(this), amount);
+            // approve connector to handle tokens depending on connector version (eg. lock or burn)
+            if (!IERC20(token).approve(zetaConnector, amount)) revert ApprovalFailed(token, zetaConnector);
+            // send tokens to connector
+            ZetaConnectorBase(zetaConnector).receiveTokens(amount);
```
[the remark this answers is not in the document]
Comment (s2imonovic) [reply contracts/evm/GatewayEVM.sol:390]: agree

File: contracts/zevm/GatewayZEVM.sol [hunk 6]
```diff
@@ -179,6 +179,13 @@ contract GatewayZEVM is
         }
     }
 
+    // TODO: Factorize common validation checks across withdraw functions.
```
Comment (lumtis) [inline contracts/zevm/GatewayZEVM.sol:182]: Can we point an issue link in the comment below TODO as well?

**Review by lumtis — APPROVED** [review]
LGTM
